File: DataStructures/Trees/treeRockPaperScissors.py
def rockPaperScissors(rounds):

  plays = ['rock', 'paper', 'scissors']
  rounds = rounds
  combos = []

  def generateCombos(rounds_to_go, played_so_far):

  	if rounds_to_go == 0:
  		combos.append(played_so_far)
  		return

  	for i in range(len(plays)):
  		current_play = plays[i]
  		played_so_far = played_so_far.append(current_play)
  		generateCombos(rounds_to_go - 1, played_so_far)

  generateCombos(rounds, [])
  print combos

rockPaperScissors(3)

File: DataStructures/Trees/treeRockPaperScissors.js
/*
* Write a function that generates every sequence of throws a single
* player could throw over a three-round game of rock-paper-scissors.
*
* Your output should look something like:
*   [["rock", "rock", "rock"],
*    ["rock", "rock", "paper"],
*    ["rock", "rock", "scissors"],
*    ["rock", "paper", "rock"],
             ...etc...
     ]
*
* Extra credit:
*   - Make your function return answers for any number of rounds.
* Example:
* rockPaperScissors(5); // => [['rock', 'rock', 'rock', 'rock', 'rock'], etc...]
*
*/

var rockPaperScissors = function (rounds) {

  var plays = ['rock', 'paper', 'scissors'];
  var rounds = rounds || 3;
  var combos = [];

  var generateCombos = function (roundsToGo, playedSoFar) {
    playedSoFar = playedSoFar || [];
    
    if (roundsToGo === 0) {
      combos.push(playedSoFar);
      return;
    }
    
    for (var i = 0; i < plays.length; i++) {
      var currentPlay = plays[i];
      generateCombos(roundsToGo - 1, playedSoFar.concat(currentPlay));
    }
    
  }

  generateCombos(rounds);
  return combos;
};

console.log(rockPaperScissors(3));
